apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/GoogleDiagnosticsWebHostBuilderExtensions.cs

```diff
@@ -0,0 +1,73 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Reflection;
+using Google.Cloud.Logging.V2;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace Google.Cloud.Diagnostics.AspNetCore
+{
+    /// <summary>
+    /// Extensions to configure Google Diagnostics on the <see cref="IWebHostBuilder"/>.
+    /// </summary>
+    public static class GoogleDiagnosticsWebHostBuilderExtensions
+    {
+        /// <summary>
+        /// Configures Google Diagostics services for the current application,
+        /// such as Logging, Tracing and Error Reporting middleware.
+        /// </summary>
+        /// <param name="builder">The <see cref="IWebHostBuilder"/> instance.</param>
+        /// <param name="projectId">
+        /// The Google Cloud Platform project ID. If unspecified and running on GAE or GCE
+        /// the project ID will be detected from the platform.
+        /// </param>
+        /// <param name="loggerOptions">Options for the logger. Can be null.</param>
+        /// <param name="loggingServiceClient">The logging client. Can be null.</param>
+        /// <param name="traceOptionsSetup">Actions to set up options for tracing services. Can be null.</param>
+        /// <param name="errorReportingOptionsSetp">Actions to set up options for error reporting services. Can be null.</param>
+        /// <returns>The <see cref="IWebHostBuilder"/> instance.</returns>
+        public static IWebHostBuilder UseGoogleDiagnostics(
+            this IWebHostBuilder builder,
+            string projectId = null,
+            LoggerOptions loggerOptions = null,
+            LoggingServiceV2Client loggingServiceClient = null,
+            Action<TraceServiceOptions> traceOptionsSetup = null,
+            Action<ErrorReportingServiceOptions> errorReportingOptionsSetp = null)
+        {
+            builder.ConfigureServices(services =>
+            {
+                services.AddSingleton<IStartupFilter>(sp =>
+                {
+                    var loggerFactory = sp.GetServiceCheckNotNull<ILoggerFactory>();
+                    return new GoogleDiagnosticsStartupFilter(loggerFactory, sp, projectId, loggerOptions, loggingServiceClient);
+                });
+
+                services.AddGoogleTrace(traceOptionsSetup ?? (options => options.ProjectId = projectId));
+                services.AddGoogleExceptionLogging(errorReportingOptionsSetp ?? (options =>
+                {
+                    options.ProjectId = projectId;
+
+                    var entryAssembly = Assembly.GetEntryAssembly();
```

Comment:
Speaking of 'good defaults'. What do you think of these? Is there any convention about what to put in these values? Right now I've configured what I usually use in my projects.

Reply:
I'm not entirely sure here @jskeet any thoughts here?

We already do some of this in the default options that default to the GAE Flex values when available.  If we do go this route I think we should move this logic into the options class.